[libc++] Deprecate `is_pod(_v)` since C++20

frederick-vs-ja opened the pull request:
Previously, commit 042f07eed8c1acba19ea04310137bee12b18045a claimed that P0767R1 was implemented in LLVM 7.0, but no deprecation warning was implemented. This patch adds the missing warnings.

frederick-vs-ja committed:
[libc++] Deprecate `is_pod(_v)` since C++20

Previously, commit 042f07eed8c1acba19ea04310137bee12b18045a claimed that
P0767R1 was implemented in LLVM 7.0, but no deprecation warning was
implemented. This patch adds the missing warnings.

frederick-vs-ja commented:
The entry for P0767R1 looks quite superfluous. Should we also change it?

https://github.com/llvm/llvm-project/blob/e42ab4c54eca0e792a0ae461481f9acbd0260363/libcxx/docs/Status/Cxx20Papers.csv?plain=1#L16

mordante commented on libcxx/docs/ReleaseNotes/21.rst:
```diff
@@ -55,6 +55,8 @@ Improvements and New Features
 Deprecations and Removals
 -------------------------
 
+- ``std::is_pod`` and ``std::is_pod_v`` are deprecated since C++20 according to the standard.
```
Can you add `P0767R1` to the Implemented paper section? And as you suggested update the C++20 paper status page too.

I feel the "according to the standard"  superfluous since that's true for most things we do. WDYT?
```suggestion
- ``std::is_pod`` and ``std::is_pod_v`` are deprecated since C++20.
```

philnik777 replied:
I think I'd rather say something like "is_pod and is_pod_v are now marked as deprecated in C++20 and later". That makes it clear that we added annotations.

mordante reviewed (commented):
Thanks for catching this!

frederick-vs-ja committed:
Merge branch 'main' into deprecate-is_pod

frederick-vs-ja committed:
Modify release notes and the status page per review comments.

mordante reviewed (approved):
LGTM, but I've some requests.

frederick-vs-ja commented on libcxx/include/type_traits:
```diff
@@ -95,8 +95,8 @@ namespace std
     // Member introspection:
-    template <class T> struct is_pod;
     template <class T> struct is_trivial;
+    template <class T> struct is_pod;          // Deprecated in C++20
```
I found that there're inconsistent `// Since C++23` comments. Should we also change them to `// C++23`?

mordante replied:
I personally feel `Since C++23` is more descriptive. Especially since we also have things like `Deprecated since C++23`, and `Removed since C++23`. Only the version seems not great at describing what the comment means.

frederick-vs-ja replied:
I decided to do the following drive-by changes:
- aligning all version comments to line 65,
- using `// since C++` instead of `// C++`, in lowercase as used in most version comments,
- changing the comments for `bool_constant` to `// since C++17` as it's a C++17 feature,
- changing the class-key of `result_of` to `struct`, which follows [N4659 [depr.meta.types]](https://timsong-cpp.github.io/cppwp/n4659/depr.meta.types) and the actual usage in libc++,
- consistently using `class` to introduce type template parameters,
- consistently using `inline constexpr` for variable templates,
- moving the comments for `is_nothrow_convertible_v` to the part for variable templates, and
- removing duplicated comments for `true_type` and `false_type`.

I found that we only use `deprecated in`/`removed in` (not "since"), so I didn't change these parts.

philnik777 replied:
This is getting a bit out of scope IMO. Half the patch isn't about deprecating `is_pod` anymore.

frederick-vs-ja replied:
Yeah. The "drive-by" changes turn out to be quite large and I want to separate them to another PR.

mordante replied:
+1 I did no realize the "drive-by" would catch so many things. Thanks a lot for doing this!
I reviewed the "drive-by" changes and I'm happy with them. Feel free to commit these as a separate NFC commit and then commit this patch. (I already approved this patch.)

frederick-vs-ja replied:
Reverted. I'd like to defer the comments cleanup after landing #128649.

frederick-vs-ja replied:
Opened #130422.

frederick-vs-ja committed:
Drive-by changes in `<type_traits>`

- aligning all version comments to line 65,
- using `// since C++` instead of `// C++`, in lowercase as used in most version comments,
- changing the comments for `bool_constant` to `// since C++17` as it's a C++17 feature,
- changing the class-key of `result_of` to `struct`, which follows N4659 [depr.meta.types] and the actual usage in libc++,
- consistently using `class` to introduce type template parameters,
- consistently using `inline constexpr` for variable templates,
- moving the comments for `is_nothrow_convertible_v` to the part for variable templates, and
- removing duplicated comments for `true_type` and `false_type`.

frederick-vs-ja committed:
Merge branch 'main' into deprecate-is_pod

frederick-vs-ja committed:
Revert "Drive-by changes in `<type_traits>`"

This reverts commit 372e8a98990879ceb85e72c9972434ffa83a4a24.

frederick-vs-ja committed:
Merge branch 'main' into deprecate-is_pod
